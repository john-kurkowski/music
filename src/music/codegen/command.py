"""Codegen command."""

from pathlib import Path

import click

from .process import main as _codegen


@click.command("codegen")
@click.argument(
    "example_audio_file",
    nargs=1,
    required=True,
    type=Path,
)
def main(example_audio_file: Path) -> None:
    """Generate code for this package.

    Requires an EXAMPLE_AUDIO_FILE from which to generate a parser.
    """
    _codegen(example_audio_file)
