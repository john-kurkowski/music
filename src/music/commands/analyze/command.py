"""Analyze command."""

import base64
from collections.abc import Iterator
from pathlib import Path

import click
import rpp  # type: ignore[import-untyped]

from music.utils import project


@click.command("analyze")
@click.argument(
    "project_dirs",
    nargs=-1,
    type=click.Path(dir_okay=True, exists=True, file_okay=True, path_type=Path),
)
def main(project_dirs: list[Path]) -> None:
    """Analyze projects for problems.

    Prints out a project's VST settings encoded in base64 for human review.
    Sometimes these contain unwanted settings, which are not possible by
    looking at a Reaper .rpp file directly.

    Accepts 0 or more PROJECT_DIRS projects. Defaults to the currently open
    project.
    """
    if not project_dirs:
        project_dirs = [Path(project.ExtendedProject().path)]

    project_fils = [
        project_dir / f"{project_dir.name}.rpp" for project_dir in project_dirs
    ]

    click.echo(
        "\n".join(
            setting
            for project_fil in project_fils
            for setting in iter_encoded_settings(project_fil)
        )
    )


def iter_encoded_settings(project_fil: Path) -> Iterator[str]:
    """Parse a Reaper project and return all VST settings encoded in base64."""
    with open(project_fil) as fil:
        parsed_project = rpp.load(fil)

    vsts = parsed_project.findall(".//VST")
    for vst in vsts:
        successful_decodes = (
            decode for child in vst.children if (decode := b64_ascii(child))
        )
        yield from successful_decodes


def b64_ascii(s: str) -> str | None:
    """Try to decode a base64 string.

    Many VST settings are encoded in Reaper as base64. Return None if it is not
    one of those strings.
    """
    try:
        return base64.b64decode(s).decode("ascii")
    except UnicodeDecodeError:
        return None
