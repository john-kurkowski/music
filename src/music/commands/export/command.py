"""Export command."""

import errno
import math
import shutil
from pathlib import Path

import click


@click.command("export")
@click.argument(
    "dst_dir",
    type=click.Path(dir_okay=True, file_okay=False, path_type=Path),
)
@click.argument(
    "files",
    nargs=-1,
    required=True,
    type=click.Path(dir_okay=False, exists=True, file_okay=True, path_type=Path),
)
def main(dst_dir: Path, files: list[Path]) -> None:
    """Export the given FILES to the given DST_DIR directory.

    Exports in album order. Maintains file metadata.
    """
    dst_dir.mkdir(exist_ok=True)

    for i, src in enumerate(files):
        dst = dst_dir / f"{i + 1:02d} - {src.with_suffix('.wav').name}"
        if not is_src_newer(src, dst):
            continue

        shutil.copyfile(src, dst)
        try:
            shutil.copystat(src, dst)
        except OSError as oserr:
            copy_success_but_missing_extended_flags = oserr.errno == errno.EINVAL
            if not copy_success_but_missing_extended_flags:
                raise


def is_src_newer(src: Path, dst: Path) -> bool:
    """Return True if the source file is newer than the destination file, or the destination file does not yet exist.

    Allow modification times within 2s of each other. Songs are very unlikely to
    take so little time to render and be this close. Without this tolerance, a
    lower resolution filesystem time will always compare different to a higher
    resolution one. `round` or `math.floor` are insufficient, as the lower
    resolution filesystem timestamps are unpredictable whether they'll go 1s up or
    down.
    """
    if not dst.exists():
        return True

    src_time = src.stat().st_mtime
    dst_time = dst.stat().st_mtime
    return src_time > dst_time and not math.isclose(src_time, dst_time, abs_tol=2.0)
