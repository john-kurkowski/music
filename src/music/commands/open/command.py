"""Open command."""

import subprocess

import click

from music.utils.project import ExtendedProject


@click.command("open")
def main() -> None:
    """Show the folder containing the current project."""
    project = ExtendedProject()
    cmd = ["open", project.path]
    subprocess.run(cmd, check=True, stderr=subprocess.PIPE)
