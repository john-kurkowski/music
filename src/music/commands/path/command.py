"""Path command."""

import click

from music.utils.project import ExtendedProject


@click.command("path")
def main() -> None:
    """Print the path to the current project."""
    project = ExtendedProject()
    click.echo(project.path)
