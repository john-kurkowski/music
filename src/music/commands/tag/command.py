"""Encode & tag command."""

from pathlib import Path

import click

from .process import main as _tag


@click.command("tag")
@click.argument(
    "file",
    required=True,
    type=click.Path(exists=True, path_type=Path),
)
def main(file: Path) -> None:
    """Encode FILE to .mp3 and tag with artist, etc metadata."""
    _tag(file)
