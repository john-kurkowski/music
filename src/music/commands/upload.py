"""Upload command."""

from pathlib import Path

import click

import music.upload
from music.render import SongVersion


@click.command("upload")
@click.argument(
    "project_dirs",
    nargs=-1,
    type=click.Path(dir_okay=True, exists=True, file_okay=False, path_type=Path),
)
@click.option(
    "--include-main",
    default=None,
    flag_value=SongVersion.MAIN,
    help=(
        "Whether to include the main version. Defaults to including all versions,"
        ' unless one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-instrumental",
    default=None,
    flag_value=SongVersion.INSTRUMENTAL,
    help=(
        "Whether to include the instrumental version."
        " Defaults to including all versions,"
        ' unless one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-acappella",
    default=None,
    flag_value=SongVersion.ACAPPELLA,
    help=(
        "Whether to include the a cappella version."
        " Defaults to including all versions, unless"
        ' one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--oauth_token",
    envvar="SOUNDCLOUD_OAUTH_TOKEN",
    help=(
        "SoundCloud OAuth token. Read from the environment variable"
        " SOUNDCLOUD_OAUTH_TOKEN."
    ),
)
def main(
    project_dirs: list[Path],
    include_main: SongVersion | None,
    include_instrumental: SongVersion | None,
    include_acappella: SongVersion | None,
    oauth_token: str,
) -> None:
    """Upload rendered output of the PROJECT_DIRS Reaper projects to SoundCloud.

    Defaults to uploading all rendered versions of the currently open project.
    """
    if not project_dirs:
        project_dirs = [Path(music.util.ExtendedProject().path)]

    versions = {
        version
        for version in (include_main, include_instrumental, include_acappella)
        if version
    } or list(SongVersion)

    files = [
        fil
        for project_dir in project_dirs
        for version in versions
        if (fil := project_dir / f"{version.name_for_project_dir(project_dir)}.wav")
        and fil.exists()
    ]

    if not files:
        raise click.UsageError("nothing to upload")

    music.upload.main(oauth_token, files)
