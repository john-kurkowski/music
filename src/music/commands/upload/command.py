"""Upload command."""

import email
from pathlib import Path

import aiohttp
import click
import rich.console
import rich.live

import music.utils
from music.utils.project import ExtendedProject
from music.utils.songversion import SongVersion

from .process import Process as UploadProcess


@click.command("upload")
@music.utils.coro
@click.argument(
    "project_dirs",
    nargs=-1,
    type=click.Path(dir_okay=True, exists=True, file_okay=False, path_type=Path),
)
@click.option(
    "--additional_headers",
    envvar="SOUNDCLOUD_ADDITIONAL_HEADERS",
    required=False,
    help=(
        "SoundCloud additional HTTP request headers. Read from the environment variable"
        " SOUNDCLOUD_ADDITIONAL_HEADERS."
    ),
)
@click.option(
    "--include-main",
    default=None,
    flag_value=SongVersion.MAIN,
    help=(
        "Whether to include the main version."
        " Defaults to including main, instrumental, and a cappella versions,"
        ' unless one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-instrumental",
    default=None,
    flag_value=SongVersion.INSTRUMENTAL,
    help=(
        "Whether to include the instrumental version."
        " Defaults to including main, instrumental, and a cappella versions,"
        ' unless one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-instrumental-dj",
    default=None,
    flag_value=SongVersion.INSTRUMENTAL_DJ,
    help=(
        "Whether to include the DJ instrumental version."
        " Defaults to including main, instrumental, and a cappella versions,"
        ' unless one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--include-acappella",
    default=None,
    flag_value=SongVersion.ACAPPELLA,
    help=(
        "Whether to include the a cappella version."
        " Defaults to including main, instrumental, and a cappella versions,"
        ' one of the "--include-*" flags is set.'
    ),
    type=SongVersion,
)
@click.option(
    "--oauth_token",
    envvar="SOUNDCLOUD_OAUTH_TOKEN",
    help=(
        "SoundCloud OAuth token. Read from the environment variable"
        " SOUNDCLOUD_OAUTH_TOKEN."
    ),
)
async def main(
    project_dirs: list[Path],
    additional_headers: str,
    include_main: SongVersion | None,
    include_instrumental: SongVersion | None,
    include_instrumental_dj: SongVersion | None,
    include_acappella: SongVersion | None,
    oauth_token: str,
) -> None:
    """Upload PROJECT_DIRS renders.

    Uploads to SoundCloud. Renders with matching name must exist in SoundCloud
    already, and will be overwritten.

    Defaults to uploading all rendered versions of the currently open project.
    """
    if not oauth_token:
        raise click.MissingParameter(
            param_hint="'SOUNDCLOUD_OAUTH_TOKEN'", param_type="envvar"
        )

    parsed_additional_headers = {**email.message_from_string(additional_headers)}

    if not project_dirs:
        project_dirs = [Path(ExtendedProject().path)]

    versions = {
        version
        for version in (
            include_main,
            include_instrumental,
            include_instrumental_dj,
            include_acappella,
        )
        if version
    } or (
        SongVersion.MAIN,
        SongVersion.INSTRUMENTAL,
        SongVersion.ACAPPELLA,
    )

    files = [
        fil
        for project_dir in project_dirs
        for version in versions
        if (fil := version.path_for_project_dir(project_dir)) and fil.is_file()
    ]

    if not files:
        click.echo("Error: nothing to upload", err=True)
        raise click.exceptions.Exit(2)

    console = rich.console.Console()
    process = UploadProcess(console)
    with rich.live.Live(process.progress, console=console, refresh_per_second=10):
        async with aiohttp.ClientSession() as client:
            uploads = await process.process(
                client, oauth_token, parsed_additional_headers, files
            )

    has_error = False
    for upload in uploads:
        if isinstance(upload, BaseException):
            has_error = True
            click.echo(upload, err=True)

    if has_error:
        raise click.exceptions.Exit(2)
