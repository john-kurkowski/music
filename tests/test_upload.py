"""Upload tests."""

import datetime
import re
from pathlib import Path
from typing import Any
from unittest import mock

import pytest
import pytest_socket  # type: ignore[import-untyped]
from click.testing import CliRunner
from syrupy.assertion import SnapshotAssertion

from music.upload.command import main as upload

from .conftest import RequestsMocks

ST_MODE_IS_FILE = 33188


@pytest.fixture
def some_paths(tmp_path: Path) -> list[Path]:
    """Create some files to upload."""
    some_paths = [
        tmp_path / "path" / "to" / "some project" / "some project.wav",
        tmp_path / "path" / "to" / "another project" / "another project.wav",
    ]
    for path in some_paths:
        path.parent.mkdir(parents=True)
        path.write_text(f".wav content for {path.stem}")

    return some_paths


def test_main_no_network_calls(some_paths: list[Path]) -> None:
    """Test that main network calls are blocked."""
    with pytest.raises(pytest_socket.SocketBlockedError):
        CliRunner(mix_stderr=False).invoke(
            upload,
            [str(path.parent.resolve()) for path in some_paths],
            catch_exceptions=False,
        )


def test_main_tracks_not_found(
    requests_mocks: RequestsMocks, snapshot: SnapshotAssertion, some_paths: list[Path]
) -> None:
    """Test main when tracks are not found/matched in the upstream database."""
    with pytest.raises(KeyError) as exc_info:
        CliRunner(mix_stderr=False).invoke(
            upload,
            [str(path.parent.resolve()) for path in some_paths],
            catch_exceptions=False,
        )

    assert exc_info.value == snapshot

    assert requests_mocks.mock_calls == snapshot


def test_main_tracks_newer(
    requests_mocks: RequestsMocks,
    snapshot: SnapshotAssertion,
    some_paths: list[Path],
) -> None:
    """Test main when tracks are newer in the upstream database, and skipped.

    Stubs enough of the local filesystem and upstream responses for timestamps
    to be checked.
    """
    old_timestamp = "2020-10-01T00:00:00Z"
    new_timestamp = "2021-10-01T00:00:00Z"

    original_stat = Path.stat

    def mock_stat(self: Path, *args: Any, **kwargs: Any) -> Any:
        """Mock `Path.stat()` to return a timestamp for some paths under test.

        Other paths use `Path.stat()`'s default implementation.
        """
        if self in some_paths:
            return mock.Mock(
                st_mode=ST_MODE_IS_FILE,
                st_mtime=datetime.datetime.fromisoformat(old_timestamp).timestamp(),
            )

        return original_stat(self, *args, **kwargs)

    def mock_get(url: str, *args: Any, **kwargs: Any) -> mock.Mock:
        if re.search(r"^https://api-v2.soundcloud.com/users/.*/tracks", url):
            return mock.Mock(
                json=mock.AsyncMock(
                    return_value={
                        "collection": [
                            {
                                "id": 1,
                                "last_modified": new_timestamp,
                                "title": "some project",
                                "permalink_url": "https://soundcloud.com/1",
                            },
                            {
                                "id": 2,
                                "last_modified": new_timestamp,
                                "title": "another project",
                                "permalink_url": "https://soundcloud.com/2",
                            },
                        ]
                    }
                )
            )

        return mock.Mock()

    requests_mocks.get.side_effect = mock_get
    with mock.patch.object(Path, "stat", autospec=True, side_effect=mock_stat):
        result = CliRunner(mix_stderr=False).invoke(
            upload,
            [str(path.parent.resolve()) for path in some_paths],
            catch_exceptions=False,
        )

    assert not result.exception
    assert result.stdout == snapshot
    assert not result.stderr

    assert requests_mocks.mock_calls == snapshot


def test_main_success(
    requests_mocks: RequestsMocks, snapshot: SnapshotAssertion, some_paths: list[Path]
) -> None:
    """Test main with multiple files succeeds.

    Stubs enough of responses for the sequence of API requests to complete.
    """
    new_timestamp = "2021-10-01T00:00:00Z"

    def mock_get(url: str, *args: Any, **kwargs: Any) -> mock.Mock:
        if re.search(r"^https://api-v2.soundcloud.com/users/.*/tracks", url):
            return mock.Mock(
                json=mock.AsyncMock(
                    return_value={
                        "collection": [
                            {
                                "id": 1,
                                "last_modified": new_timestamp,
                                "title": "some project",
                                "permalink_url": "https://soundcloud.com/1",
                            },
                            {
                                "id": 2,
                                "last_modified": new_timestamp,
                                "title": "another project",
                                "permalink_url": "https://soundcloud.com/2",
                            },
                        ]
                    }
                )
            )
        elif re.search(
            r"^https://api-v2.soundcloud.com/uploads/.*/track-transcoding", url
        ):
            return mock.Mock(
                json=mock.AsyncMock(
                    return_value={
                        "status": "finished",
                    }
                )
            )

        return mock.Mock()

    def mock_post(url: str, *args: Any, **kwargs: Any) -> mock.Mock:
        if re.search(
            r"^https://api-v2.soundcloud.com/uploads/track-upload-policy", url
        ):
            return mock.Mock(
                json=mock.AsyncMock(
                    return_value={
                        "headers": {"some-uploader-id": "some-uploader-value"},
                        "url": "https://some-url",
                        "uid": "stub-uid",
                    }
                )
            )

        return mock.Mock()

    requests_mocks.get.side_effect = mock_get
    requests_mocks.post.side_effect = mock_post

    result = CliRunner(mix_stderr=False).invoke(
        upload,
        [str(path.parent.resolve()) for path in some_paths],
        catch_exceptions=False,
    )

    assert not result.exception
    assert result.stdout == snapshot
    assert not result.stderr

    assert requests_mocks.mock_calls == snapshot
