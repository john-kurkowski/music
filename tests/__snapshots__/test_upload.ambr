# serializer version: 1
# name: test_main_success
  '''
  some project.wav: https://soundcloud.com/1
  another project.wav: https://soundcloud.com/2
  ⠋ Uploading "some project.wav"    0:00:00                             0/29 bytes
  ⠋ Uploading "another project.wav" 0:00:00                             0/32 bytes
    Transcoding "some project.wav"    0:00:00
    Transcoding "another project.wav" 0:00:00
  '''
# ---
# name: test_main_success.1
  dict({
    'get': _CallList([
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/users/41506/tracks',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'params': dict({
            'limit': 999,
          }),
          'timeout': 10,
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/stub-uid/track-transcoding',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/stub-uid/track-transcoding',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
        }),
      ),
    ]),
    'post': _CallList([
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/track-upload-policy',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'json': dict({
            'filename': 'some project.wav',
            'filesize': 29,
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/stub-uid/track-transcoding',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/track-upload-policy',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'json': dict({
            'filename': 'another project.wav',
            'filesize': 32,
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/uploads/stub-uid/track-transcoding',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
        }),
      ),
    ]),
    'put': _CallList([
      _Call(
        '',
        tuple(
          'https://some-url',
        ),
        dict({
          'data': 'ASYNC_ITERABLE_HERE',
          'headers': dict({
            'some-uploader-id': 'some-uploader-value',
          }),
          'timeout': 600,
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/tracks/soundcloud:tracks:1',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'json': dict({
            'track': dict({
              'replacing_original_filename': 'some project.wav',
              'replacing_uid': 'stub-uid',
              'title': 'some project',
            }),
          }),
        }),
      ),
      _Call(
        '',
        tuple(
          'https://some-url',
        ),
        dict({
          'data': 'ASYNC_ITERABLE_HERE',
          'headers': dict({
            'some-uploader-id': 'some-uploader-value',
          }),
          'timeout': 600,
        }),
      ),
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/tracks/soundcloud:tracks:2',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'json': dict({
            'track': dict({
              'replacing_original_filename': 'another project.wav',
              'replacing_uid': 'stub-uid',
              'title': 'another project',
            }),
          }),
        }),
      ),
    ]),
  })
# ---
# name: test_main_tracks_newer
  '''
  Skipping already uploaded "some project.wav"
  Skipping already uploaded "another project.wav"
  
  '''
# ---
# name: test_main_tracks_newer.1
  dict({
    'get': _CallList([
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/users/41506/tracks',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'params': dict({
            'limit': 999,
          }),
          'timeout': 10,
        }),
      ),
    ]),
    'post': _CallList([
    ]),
    'put': _CallList([
    ]),
  })
# ---
# name: test_main_tracks_not_found
  KeyError("Tracks to upload not found in SoundCloud: ['another project', 'some project']")
# ---
# name: test_main_tracks_not_found.1
  dict({
    'get': _CallList([
      _Call(
        '',
        tuple(
          'https://api-v2.soundcloud.com/users/41506/tracks',
        ),
        dict({
          'headers': dict({
            'Accept': 'application/json',
            'Authorization': 'OAuth stub-fake-token',
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36',
          }),
          'params': dict({
            'limit': 999,
          }),
          'timeout': 10,
        }),
      ),
    ]),
    'post': _CallList([
    ]),
    'put': _CallList([
    ]),
  })
# ---
